packages/flutter/lib/src/rendering/object.dart

```diff
@@ -671,6 +672,105 @@ class _ForkingSemanticsFragment extends _SemanticsFragment {
+  void flushLayout() {
+    Timeline.startSync('Layout');
+    _debugDoingLayout = true;
+    try {
+      // TODO(ianh): assert that we're not allowing previously dirty nodes to redirty themeselves
+      while (_nodesNeedingLayout.isNotEmpty) {
+        List<RenderObject> dirtyNodes = _nodesNeedingLayout;
+        _nodesNeedingLayout = <RenderObject>[];
+        for (RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => a.depth - b.depth)) {
+          if (node._needsLayout && node.owner == this)
+            node._layoutWithoutResize();
+        }
+      }
+    } finally {
+      _debugDoingLayout = false;
+      Timeline.finishSync();
+    }
+  }
+
+  List<RenderObject> _nodesNeedingCompositingBitsUpdate = <RenderObject>[];
+  /// Updates the [needsCompositing] bits.
+  ///
+  /// Called as part of the rendering pipeline after [flushLayout] and before
+  /// [flushPaint].
+  void flushCompositingBits() {
+    Timeline.startSync('Compositing Bits');
+    _nodesNeedingCompositingBitsUpdate.sort((RenderObject a, RenderObject b) => a.depth - b.depth);
+    for (RenderObject node in _nodesNeedingCompositingBitsUpdate) {
+      if (node.owner == this)
+        node._updateCompositingBits();
+    }
+    _nodesNeedingCompositingBitsUpdate.clear();
+    Timeline.finishSync();
+  }
+
+  List<RenderObject> _nodesNeedingPaint = <RenderObject>[];
+  bool _debugDoingPaint = false;
+  bool get debugDoingPaint => _debugDoingPaint;
+  /// Update the display lists for all render objects.
+  ///
+  /// This function is one of the core stages of the rendering pipeline.
+  /// Painting occurs after layout and before the scene is recomposited so that
+  /// scene is composited with up-to-date display lists for every render object.
+  ///
+  /// See [FlutterBinding] for an example of how this function is used.
+  void flushPaint() {
+    Timeline.startSync('Paint');
+    _debugDoingPaint = true;
+    try {
+      List<RenderObject> dirtyNodes = _nodesNeedingPaint;
+      _nodesNeedingPaint = <RenderObject>[];
+      // Sort the dirty nodes in reverse order (deepest first).
+      for (RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => b.depth - a.depth)) {
+        assert(node._needsPaint);
+        if (node.owner == this)
+          PaintingContext.repaintCompositedChild(node);
+      };
+      assert(_nodesNeedingPaint.length == 0);
+    } finally {
+      _debugDoingPaint = false;
+      Timeline.finishSync();
+    }
+  }
+
+  bool _semanticsEnabled = false;
+  bool _debugDoingSemantics = false;
+  List<RenderObject> _nodesNeedingSemantics = <RenderObject>[];
+
+  void flushSemantics() {
+    Timeline.startSync('Semantics');
+    assert(_semanticsEnabled);
+    assert(() { _debugDoingSemantics = true; return true; });
+    try {
+      _nodesNeedingSemantics.sort((RenderObject a, RenderObject b) => a.depth - b.depth);
+      for (RenderObject node in _nodesNeedingSemantics) {
+        if (node._needsSemanticsUpdate)
```

Comment:
Do we want to check for `node.owner == this` like we do for the other stages of the pipeline?